src/FsCodec.SystemTextJson/UnionConverter.fs

```diff
@@ -1,10 +1,26 @@
 ﻿namespace FsCodec.SystemTextJson
 
+open FsCodec.SystemTextJson
+open FsCodec.SystemTextJson.Core
 open FSharp.Reflection
 open System
+open System.Reflection
+open System.Text.Json
+
+/// <summary>Use this attribute in combination with a JsonConverter/UnionConverter attribute to specify
+/// your own name for a discriminator and/or a catch-all case for a specific discriminated union.
+/// If this attribute is set, its values take precedence over the values set on the converter itself.
+/// E.g. <c>[<JsonConverter(typeof<UnionConverter>); JsonUnionConverterOptions("type")>]</c></summary>
+/// <remarks>Not inherited because JsonConverters don't get inherited right now.
+/// https://github.com/dotnet/runtime/issues/30427#issuecomment-610080138</remarks>
+[<AttributeUsage(AttributeTargets.Class ||| AttributeTargets.Struct, AllowMultiple = false, Inherited = false)>] 
+type JsonUnionConverterOptionsAttribute(discriminator : string) =
```

Comment:
any reason to make this mandatory ? the catchall is probably more likely to be customized?

Reply:
This wall of text relates to a couple of items of feedback, and I realized caused me a bit of churn in my commits, but this is the piece of feedback that made me realized I needed to articulate my thoughts so I've placed it here.

---

I think I was requiring the discriminator to be specified as a way to solve an issue I was having about inferring the user's intentions with these annotations and the outcome they'd expect.

## Context
Let's imagine the discriminator is _not_ required right now, both `Discriminator` and `CatchAllCase` are both nullable on `JsonUnionConverterOptionsAttribute`, a user can set one, both, or none. We also have default options set on the converter, "case" as a discriminator and no catch-all case.

I'll use two imaginary DUs here called `DiscriminatorOption` and `CatchAllCaseOption` to represent the possible user intents when they are setting options via an attribute:
```fsharp
type DiscriminatorOption =
| CustomDiscriminator of string
| DefaultDiscriminator

type CatchAllCaseOption =
| CustomCase of string
| NoCase
| Default
```

## Expected behaviour
Inferring a user's intended `DiscriminatorOption` when they set the `Discriminator` of the attribute is fine as it's binary and quite understandable when represented with null/not-null:
```
[JsonUnionConverterOptions(Discriminator = "type")]
= DiscriminatorOption.CustomDiscriminator "type"
= "I want to set a specific discriminator for this union"
```
```
[JsonUnionConverterOptions()]
= DiscriminatorOption.DefaultDiscriminator
= "I want to use whatever the default discriminator is."
```

However `CatchAllCaseOption` is ternary and I don't know what users would expect:
```
[JsonUnionConverterOptions(CatchAllCase = "other")]
= CatchAllCaseOption.CustomCase "other"
= "I want to set a specific catch-all case for this union"
```
```
[JsonUnionConverterOptions()]
= ...? Does this mean they want no catch-all case or that they want to use the default catch-all case option
```

## Attempted solution
My attempted solution for that was making the discriminator mandatory so it was explicit users were choosing custom options.
If they set the attribute they are saying `DiscriminatorOption = CustomDiscriminator` and `CatchAllCaseOption = CustomCase | NoCase`; conversely 'DefaultDiscriminator' and 'DefaultCase' were selectable by _not_ having the attribute at all.

However you raise a good point that the use case for this attribute is mainly setting the `CatchAllCaseOption` (and users will probably be happy with the default discriminator).

## Potential solutions
Here's are a couple of options I thought of to try to address understanding the user's intent but also that use case:

1. Split these options into two attributes: `JsonUnionConverterDiscriminatorAttribute`; `JsonUnionConverterCatchAllCaseAttribute`
   
   ```
   [JsonUnionConverterDiscriminator("type")]
   = DiscriminatorOption.CustomDiscriminator "type"
   = "I want to set a specific discriminator for this union"
   ```
   ```
   (*no JsonUnionConverterDiscriminator attribute*)
   = DiscriminatorOption.DefaultDiscriminator
   = "I want to use whatever the default discriminator is."`
   ```

   ```
   [JsonUnionConverterCatchAllCase(CaseName = "other")]
   = CatchAllCaseOption.CustomCase "other"
   = "I want to set a specific catch-all case for this union"
   ```
   ```
   [JsonUnionConverterCatchAllCase()]
   = CatchAllCaseOption.NoCase
   = "I want no catch-all case for this union"
   ```
   ```
   (*no JsonUnionConverterCatchAllCase attribute*)
   = CatchAllCaseOption.Default
   = "I want the default catch-all case for this union, if there is one"
   ```

2. Stick with `JsonUnionConverterOptionsAttribute` but make the discriminator not required, and infer a null `CatchAllCase` to mean `CatchAllCaseOption.NoCase`, overriding whatever is the default option (which as it happens, is also no catch-all case).
   If this is done, I think I also ought to remove the `catchAllCase` constructor parameter from `UnionConverter` so users _can't_ set a catchAllCase if they are manually creating and registering a converter via `Options.Create`. This is because if they then set wanted to set a custom discriminator via an attribute, this solution would now infer it as them also choosing `CatchAllCaseOption.NoCase` (which would override what they set in the constructor of the converter which I think would be unexpected behaviour).

## Next steps
If you have a strong feeling about those options, spot flaws in my reasoning, or better suggestions please let me know, otherwise I'll go ahead an implement the solution described in `1` in the next couple of days.

Reply:
A catch-all is not the common case. As with TypeSafeEnumConverter, simply rejecting by throwing is fine.

There never was the notion of a 'default catch-all' (it was added as a feature later, and e.g. for that to become a default would be annoying/breaking).

I thus prefer a single options attribute, and removing the notion of `CatchAllCaseOption.Default`

Having said all that, it's up to you - my default stance as an implementer is to work toward an easy to scan impl and easy to test (meaning easy to define) spec (which I think matches the preceding - KISS principle ;)).

Reply:
🤔... unless the default catch-all was defined as the single case with a body that's a `JsonElement`, but thats a pretty loose approach not in alignment with STJ design in general, I'd say ;)

I think with these things, in practice, explicit is better than implicit - the default mode of operation is that a) unions dont get converted b) applying the JsonUnionConverter means they get converted a single way (using `"case"`) with no default case
 c) only by applying the JsonUnionConverterOptions would you every end up with catch-all semantics

Reply:
Okay, excellent. That works for me.

I will stick with a single options attribute and remove the notion of `CatchAllCaseOption.Default` by removing `catchAllCase` from the `UnionConverter` constructor.

Reply:
👍what about if there was an arg in the ctor (with `null` signifying no catch-all), and:
- the factory did the (only) reflection lookup to find options
- the converter can validate that the designated catch-all exists on construction (covers case where you are registering the converter by adding directly to the Options's Converters list, which lets you manage a type without having to apply attributes to it for strange corner cases)

(I hope I'm not missing lots of reasons why it can't be like this!)

Reply:
RE: Catch-all in ctor arg

I think this has the issue around understanding the user's intent

What if they have registered a converter and specified a catch-all case via the ctor,
```fsharp
Options.Create(UnionConverter<MyDU>((*discriminator*) "type", (*catchAllCase*) "other")
```

and then use the attribute to specify a custom discriminator (but left out the catch-all case in said attribute)?
```fsharp
[JsonUnionConverterOptions(Discriminator = "kind")]
```

Would you (the user) expect the converter to use the catch-all case as specified in the ctor, or for it to not use a catch-all case?

And then what would the user be trying to do if they did this?
```fsharp
[JsonUnionConverterOptions(CatchAllCase = null)]
```
(I know we, the converter, can't differentiate that from the previous example, but in terms of user intent...)

Reply:
>If this attribute is set, its values take precedence over the values set on the converter via its constructor.

not getting why this needs to be ?

I thought I/the user would only expect the JsonUnionConverterOptionsAttribute to come into play if I'd asked for the conversion via JsonConverterAttribute; if I programmatically configure it by calling a ctor with 2 defaulting args (with xmldoc saying this is the choice, attributes are not looked up)

i.e. the default ctor could do the lookup and then pass the opts to the primary constructor: 
```
type JsonUnionConverter<'T>(discriminator, ?catchAll) =
    ...
    do match catchAll with
          | None -> ()
          | Some ca ->
              <throw if invalid case nominated>
    new() =
        let d, ca = <attribute lookup on T>, default to "case" for d
        JsonUnionConverter(discriminator=d, ?catchAll=ca) 
```

Reply:
Ohhh okay, that was our misalignment. I had assumed you/the user might use `JsonUnionConverterOptionsAttribute` without `JsonConverterAttribute`.
('might', I merely mean we can't stop someone applying the options attribute even though they're registering the converter themselves, even though it might not be a great idea.)

I feel like I understand your position now. I'm happy to implement so we only fetch the attributes there are no ctor arguments supplied. (And yes, validating the nominated case is a good idea.)

Reply:
Cool, sorry for any confusion caused to steer you into this direction (I'm extremely lazy and would never even have considered going to the lengths of completeness you were considering; a lack of imagination on my part!)

I think ultimately this should yield a compact impl that one can scan and grok without referring to this whole dialogue in the end!

Reply:
Compact impl sounds good